### PR TITLE
Intrinsics

### PR DESCRIPTION
**Pull request by ZuseZ4**
fixes #703
fixes #802

@wsmoses I'll dislike the type part and will most likely adjust it once I start moving the mem related intrinsics and see what works better. For the meantime it should work. 
Also need to add some lines to automatically collect the useful attributes from the intrinsics and apply them to the outer call,
but I'd add that later if you don't mind.
Will add a test for the remainder in a sec.

### CODE REVIEW

File: enzyme/Enzyme/InstructionDerivatives.td [hunk 1]
```diff
@@ -110,9 +115,22 @@ def : CallPattern<(Op $x),
...
+                  [(FDiv (DiffeRet<"">), (Intrinsic<"sqrt", ["args[0]->getType()"]> (FSub (ConstantFP<"1.0"> $x), (FMul $x, $x)))  )]                  
+                  >;
+def : CallPattern<(Op $x),
+                  ["log1p", "log1pf", "log1pl"],
+                  [(FDiv (DiffeRet<"">), (FAdd (ConstantFP<"1.0"> $x), $x))]
+                  >;
 def : CallPattern<(Op $x),
                   ["expm1"],
-                  [(FMul (FAdd (ConstantFP<"1.0"> $x), (Call<(SameFunc), [ReadNone,NoUnwind]> $x) ), (DiffeRet<"">))]
+                  [(FMul (Intrinsic<"exp", ["args[0]->getType()"]> $x), (DiffeRet<"">))]
```
Comment (wsmoses) [inline enzyme/Enzyme/InstructionDerivatives.td:135]: Can you make a typeof expr, like maybe (Typeof $x)

File: enzyme/test/Enzyme/ForwardMode/remainder.ll [hunk 2]
```diff
@@ -0,0 +1,39 @@
+; RUN: %opt < %s %loadEnzyme -enzyme -enzyme-preopt=false -O3 -S | FileCheck %s
```
Comment (wsmoses) [inline enzyme/test/Enzyme/ForwardMode/remainder.ll:1]: Can you change this to only do mem2reg simplifycfg and instsimplify?

We should check that the differeturn is handled properly

File: enzyme/Enzyme/InstructionDerivatives.td (hunk 1 above)
Comment (wsmoses) [reply enzyme/Enzyme/InstructionDerivatives.td:135]: Alternatively maybe something like [0,2] denotes the types of the 0th op and the 2nd op

Comment (ZuseZ4) [reply enzyme/Enzyme/InstructionDerivatives.td:135]: I think I'll go for the TypeOf version for now, imho more readable.

File: enzyme/test/Enzyme/ForwardMode/remainder.ll (hunk 2 above)
Comment (ZuseZ4) [reply enzyme/test/Enzyme/ForwardMode/remainder.ll:1]: Done, output looks correct to me.

File: enzyme/test/Enzyme/ReverseMode/remainder.ll [hunk 3]
```diff
@@ -0,0 +1,38 @@
+; RUN: %opt < %s %loadEnzyme -enzyme -enzyme-preopt=false -O3 -S | FileCheck %s
```
Comment (wsmoses) [inline enzyme/test/Enzyme/ReverseMode/remainder.ll:1]: This and the above tests still don't seem updated?

File: enzyme/Enzyme/InstructionDerivatives.td [hunk 4]
```diff
@@ -110,9 +115,22 @@ def : CallPattern<(Op $x),
 def : CallPattern<(Op $x),
                   ["tan", "tanf", "tanl"],
                   [(FMul (DiffeRet<"">), (FAdd (ConstantFP<"1.0"> $x), (FMul(Call<(SameFunc), [ReadNone,NoUnwind]> $x):$c, $c)))]>;
-
-// TODO replace with intrinsic call to exp
+def : CallPattern<(Op $x, $y),
+                  ["remainder"],
+                  [
+                    (DiffeRet<"">),
+                    (FMul (DiffeRet<"">), (Intrinsic<"round", ["orig->getOperand(0)->getType()"]> (FDiv $x, $y)))
```
Comment (wsmoses) [inline enzyme/Enzyme/InstructionDerivatives.td:122]: I think this should be -dret * round(x/y).  that is to say -- shouldn't there be a minus sign here?

File: enzyme/test/Enzyme/ReverseMode/remainder.ll (hunk 3 above)
Comment (ZuseZ4) [reply enzyme/test/Enzyme/ReverseMode/remainder.ll:1]: Have it locally. I didn't finish the other update yet and I generally try to push multiple commits simultaneously to not wait so long for CI.

File: enzyme/Enzyme/InstructionDerivatives.td (hunk 1 above)
Comment (ZuseZ4) [reply enzyme/Enzyme/InstructionDerivatives.td]: Done. This way I can add something else impl TypeInfo for the mem things.

File: enzyme/Enzyme/InstructionDerivatives.td (hunk 4 above)
Comment (ZuseZ4) [reply enzyme/Enzyme/InstructionDerivatives.td]: good catch, lost it when replacing constantFP<"-1"> by differet

**Pull request merged**